src/core/iop/cso_reader.cpp

```diff
@@ -0,0 +1,349 @@
+    assert(m_virtptr + size <= m_size);
+    
+    const uint64_t start = m_virtptr;
+    const uint64_t end = start + size;
+    const auto start_block = (uint32_t)(start / m_blocksize);
+    const auto end_block = (uint32_t)(end / m_blocksize);
+    
+    uint64_t total_read = 0;
+    for (uint32_t i = start_block; i <= end_block; ++i)
+    {
+        if (!read_block_internal(i))
+            return total_read;
+        
+        const uint64_t local_ofs = start - start_block * m_blocksize;
+        uint64_t readlen = m_blocksize;
+        if (i == start_block)
+            readlen -= local_ofs;
+        if (i == end_block)
+            readlen -= m_blocksize - (end - end_block * m_blocksize);
+        
+        memcpy(dst, m_frame + local_ofs, readlen);
+        total_read += readlen;
+        m_virtptr += readlen;
+        dst += readlen;
+    }
+    
+    return total_read;
+}
+
+
+bool CSO_Reader::open(const char* path)
+{
+    close();
+    m_file = std::ifstream(path, std::ios::binary | std::ios::ate);
+    if (!m_file.is_open())
+    {
+        fprintf(stderr, "failed to open file\n");
+        return false;
+    }
+    auto file_len = m_file.tellg();
+    m_file.seekg(0, std::ios::beg);
+    
+    CSO_Header header;
+    m_file.read((char*)&header.magic, sizeof(uint32_t));
+    m_file.read((char*)&header.header_len, sizeof(uint32_t));
+    m_file.read((char*)&header.raw_len, sizeof(uint64_t));
+    m_file.read((char*)&header.block_len, sizeof(uint32_t));
+    m_file.read((char*)&header.version, sizeof(uint8_t));
+    m_file.read((char*)&header.index_shift, sizeof(uint8_t));
+    m_file.read((char*)header.reserved, 2 * sizeof(uint8_t));
+    
+    // validate header
+    if (header.magic != FOURCC("CISO"))
+    {
+        fprintf(stderr, "file is not a CSO!\n");
+        return false;
+    }
+    if (header.version > 1)
+    {
+        fprintf(stderr, "unsupported CSO version or corrupt file\n");
+        return false;
+    }
+    
+    // reinstate these if CSOv2 support is ever added
+    /*
+    if (header.version == 2 && header.header_len != 0x18)
+        return false;
+    if (header.version == 2 && (header.reserved[0] || header.reserved[1]))
+        return false;
+    */
+    
+    // read indices
+    auto num_entries = (uint32_t)((header.raw_len + header.block_len - 1) / header.block_len) + 1;
+    m_indices = new uint32_t[num_entries];
+    m_file.read((char*)m_indices, num_entries * sizeof(uint32_t));
+    if ((uint32_t)m_file.gcount() != num_entries * sizeof(uint32_t))
+    {
+        fprintf(stderr, "failed to read CSO indices\n");
+        close();
+        return false;
+    }
+    
+    // sanity check indices
+    uint32_t lastidx = m_indices[0];
+    if ((lastidx & ~IDX_COMPRESS_BIT) << header.index_shift < 0x18)
+    {
+        fprintf(stderr, "CSO indices are corrupted (starts within header)\n");
+        close();
+        return false;
+    }
+    
+    uint32_t framesize = header.block_len + (1 << m_shift);
```

Comment:
`m_shift` hasn't been set yet at this point, so it's always 0 right?  Is that intentional?  Currently, this probably means very large ISOs will falsely fail the sanity check.

Also: if `index_shift` is 0, no block should theoretically be larger than `block_len`.  The maximal amount of padding ought to be `(1 << index_shift) - 1`.  That being said, this is only true of an optimal CSO file (which maxcso generally produces.)  Some CSO tools exist which produce files that are not optimal, but then again these tools also don't handle >2GB properly in many cases.

Final comment: be careful about calling this format "CISO".  There's a completely unrelated format used for Wii discs called CISO (sometimes referred to as CSO.)  In this case, the format is always called CSO, though the file header has CISO.  I know... it's confusing.

-[Unknown]

Reply:
It's been a while since I last touched this code but that indeed looks like a typo, and should be `header.index_shift` like the code around it. (I can't think of why that would be intentional, probably just an oversight after moving some lines around on my part)

So far to my knowledge people (myself included) have only been testing with maxcso-compressed CSOs, I've yet to run into older tools or anyone using them. Still in the spirit of being correct I've tried to stick to the spec and not make any assumptions that could be broken, ideally those sub-optimal files should be readable too.

Yes that was confusing (and also part of why over half the search engine results are irrelevant), the only reference to the format being called "CISO" in the code itself is a comment up the top of the cpp file, which we could change easily.